### ISSUE TITLE
Epic: first-class banded ridge encoding with per-response tuning and scalable solvers

### ISSUE THREAD

**Issue by bbuchsbaum** [bbuchsbaum/rMVPA#71]
# Epic: first-class banded ridge encoding with per-response tuning and scalable solvers

> **Status:** complete; release-certified at `5529a69a10fc1429d18c4e1318daae4325f81c17`  
> **Empirical motivation:** #70  
> **Primary outcome:** a supported, leakage-safe banded-ridge encoding workflow for voxelwise and other multi-response neuroimaging data
> **Local tracker:** Mote epic `bd-01M12V25A892F2PDWMNZQQD6GB`; BR1-BR6 closed; epic closed at `5529a69a10fc1429d18c4e1318daae4325f81c17`

## Decision

rMVPA is the home for the **neuroimaging-facing model and workflow**: feature-set design, blocked cross-validation, chunked voxelwise execution, spatial outputs, diagnostics, and eventual reuse by domain-adaptation models.

The numerical implementation will have a **pure matrix core** that does not depend on `mvpa_dataset`, ROI, searchlight, or spatial classes. This keeps the mathematical contract independently testable and leaves the core extractable if it later warrants a small standalone package.

This is an epic, not a single patch. The work will land as independently reviewable child issues in the dependency order below.

## Why this belongs in rMVPA

rMVPA already has the relevant integration surface:

- `feature_sets()`, `blocks()`, and `by_set()` represent grouped predictors;
- `feature_sets_design()` carries continuous stimulus-to-brain designs;
- blocked/custom cross-validation objects express run- and segment-aware splits;
- regional/global execution and spatial result infrastructure can emit neuroimaging outputs;
- `banded_ridge_da_model()` contains a correct fixed-penalty generalized-ridge solve that can become a differential reference during extraction.

The missing capability is not merely a public wrapper. rMVPA currently has **no band-weight search or per-response hyperparameter-selection engine**. The new model therefore requires a reference core, nested tuning, a voxelwise execution path, optimized primal/dual backends, and explicit result semantics.

## Scientific contract

Given feature spaces

$$
X = [X_1\;X_2\;\cdots\;X_B], \qquad Y \in \mathbb{R}^{n \times V},
$$

fit each response using

$$
\hat\beta = \arg\min_\beta
\left\|y - \sum_{b=1}^B X_b\beta_b\right\|_2^2
+ \sum_{b=1}^B \lambda_b\|\beta_b\|_2^2.
$$

For tuning, parameterize

$$
\lambda_b = \alpha/\theta_b,
\qquad \theta_b \ge 0,
\qquad \sum_b \theta_b = 1,
$$

where `alpha` controls overall shrinkage and `theta` controls its allocation across feature spaces. A zero `theta_b` is treated as the limiting case that excludes a band, not as a literal division by zero.

The implementation must preserve these estimands:

1. **Selection estimand:** inner-CV predictive loss for each response. Default: mean squared error. Correlation or $R^2$ may be explicit alternatives but must not silently replace the default.
2. **Performance estimand:** metrics computed once from concatenated outer out-of-fold predictions for each response, not an unweighted mean of fold-level $R^2$ values.
3. **Attribution estimand:** leave-one-feature-set-out predictive $\Delta R^2_b = R^2_{\text{full}} - R^2_{-b}$, with the full and reduced models independently retuned inside each outer-training set.
4. **No generic shared-variance claim:** leave-one-set-out effects may be negative, their sum is unconstrained relative to full $R^2$, and they do not form an additive decomposition when feature spaces are correlated.

The simulation in #70 is motivation, not release proof. Its primary reported contrast should be the absolute correlation change (`+0.118` at the stated intermediate SNR), with seeds, folds, candidate sets, and uncertainty made reproducible before it is cited in package documentation.

### Pure computational core

The core will accept numeric matrices/list-of-matrices, response matrices, explicit train/validation indices, and validated hyperparameter candidates. It will own:

- fold-local centering/scaling and intercept restoration;
- fixed-penalty fitting and prediction;
- candidate generation and deterministic random search over `theta`;
- per-response selection of `alpha` and `theta`;
- primal-direct, primal-SVD, and dual-kernel backends;
- target batching and optional refitting of primal or dual weights;
- solver, search, preprocessing, and fold provenance.

It will not know about voxels, masks, ROIs, searchlights, or neuroimaging file formats.

### rMVPA integration layer

The rMVPA model will own:

- validation of `mvpa_dataset` and `feature_sets_design` alignment;
- explicit outer and inner cross-validation specifications;
- blocked/purged time-series folds;
- chunked execution over responses without overlapping-searchlight refits;
- per-response maps for out-of-fold metrics and selected hyperparameters;
- optional out-of-fold predictions and optional/on-disk weights;
- stable output schemas and user-facing diagnostics.

`banded_ridge_da_model()` will retain its current public behavior. It may adopt the shared low-level fixed-penalty core only after a dedicated parity child issue proves unchanged results.

## Proposed API direction

The final signature will be settled after the reference-core prototype, but the public contract should resemble:

```r
banded_ridge_model(
  dataset,
  design,                         # owns feature_sets and grouping
  crossval,                       # outer performance folds
  tune_crossval = NULL,           # inner folds, derived safely when possible
  alpha_grid = 10^seq(0, 6, length.out = 13),
  band_search = c("random", "grid", "fixed"),
  n_search = 30L,
  alpha_scope = c("response", "roi"),
  theta_scope = c("response", "roi", "fixed"),
  selection_metric = c("mse", "r2", "correlation"),
  solver = c("auto", "primal", "dual"),
  target_batch_size = NULL,
  compute_delta_r2 = FALSE,
  delta_sets = NULL,
  return_predictions = FALSE,
  return_weights = c("none", "dual", "primal")
)
```

Contract decisions:

- `design`, not the model constructor, owns the feature-set specification and block metadata.
- `alpha_scope` and `theta_scope` are separate so “per-response regularisation” is unambiguous.
- Time-series analyses must supply block-aware CV or an explicit custom split; the model will not silently choose random folds.
- All preprocessing is fitted on the applicable training rows and applied to validation/test rows.
- The penalty scale is documented against the exact objective above; no hidden division by `n`.
- `solver = "auto"` is enabled only after benchmark evidence defines a stable time/memory heuristic.
- Weights are opt-in because a dense $p \times V$ coefficient matrix can be much larger than the fitted data.

### 1. Contract and direct reference core

Deliver:

- a pure internal representation of grouped predictors and validated `alpha`/`theta` candidates;
- fold-local standardization and response centering;
- a small, direct primal solver using the generalized-ridge normal equations;
- prediction and optional coefficient recovery;
- explicit validation for shapes, finite data, non-negative penalties, empty bands, constant columns/responses, and rank deficiency.

Accept when:

- the solver agrees with an independent direct analytic implementation under condition-scaled absolute and relative tolerances;
- equal band penalties reproduce ordinary ridge;
- feature-column permutation within a band and band-order relabeling preserve predictions;
- the current `alpha_recall = 0` stacked solve agrees on matched standardized inputs;
- focused tests include typical, rank-deficient, ill-conditioned, and extreme-penalty cases.

### 2. Nested per-response tuning

Deliver:

- deterministic candidate generation over the `theta` simplex;
- an `alpha` path for every fixed `theta`;
- inner-fold scores retained per candidate and response;
- independent `alpha_scope` and `theta_scope` semantics;
- refitting on the complete outer-training set using only inner-selected hyperparameters;
- reproducible batching over responses.

Accept when:

- mutating outer-test responses cannot change selected hyperparameters, fitted preprocessing, or outer-training predictions;
- each response can choose a different `alpha` and `theta` on a constructed fixture;
- fixed/shared scopes reproduce their corresponding constrained reference searches;
- pure-noise repeated simulations are calibrated around chance, but are not the sole leakage test;
- selection is deterministic for a fixed seed, candidate list, folds, and batch size.

### 3. Single-domain rMVPA model and voxelwise execution

Deliver:

- `banded_ridge_model()` using `feature_sets_design()` and rMVPA cross-validation objects;
- a chunked whole-brain/per-response execution path;
- out-of-fold MSE, correlation, and $R^2$ maps plus selected-`alpha` and band-weight maps;
- optional prediction retention and explicit coefficient-storage policy;
- actionable errors for unsafe time-series CV.

Accept when:

- a public end-to-end example runs without private `:::` calls;
- concatenated out-of-fold predictions reproduce reported per-response metrics;
- changing response chunk size leaves results invariant within numerical tolerance;
- regional batching and whole-brain batching agree for the same non-overlapping response sets;
- overlapping searchlight execution is unsupported in v1 rather than producing ambiguous maps.

### 4. Optimized primal-SVD and dual-kernel backends

Deliver:

- an SVD path that reuses a decomposition across the `alpha` path for a fixed `theta` and training fold;
- per-band train kernels and correctly centered test/train cross-kernels;
- dual fitting, prediction, and optional primal/dual weight recovery;
- target and alpha batching with bounded memory;
- benchmark-derived `solver = "auto"` behavior.

Complexity claims must count search candidates and folds. Changing `theta` changes the rescaled design/kernel and therefore generally requires a new decomposition for each candidate and training fold; only the `alpha` path is cheap after that decomposition.

Accept when:

- direct-primal, SVD-primal, and dual predictions agree across both $p<n$ and $p>n$ fixtures;
- equivalence holds with intercepts, fold-local standardization, unequal bands, and near-singular inputs;
- optional differential tests agree with a pinned external oracle such as `himalaya` on exported fixtures;
- representative benchmarks record runtime and peak memory for small-$p$ and large-$p$ regimes;
- `auto` never selects a backend whose estimated kernel or coefficient storage exceeds the configured memory contract.

### 5. Predictive feature-set attribution

Deliver:

- leave-one-set-out predictive $\Delta R^2$ for explicitly requested sets;
- nested retuning of every reduced model within outer-training data;
- per-response outputs and computation diagnostics;
- clear naming that distinguishes predictive drop-out effects from commonality analysis or additive decompositions.

Accept when:

- hand-constructed correlated/suppressor fixtures demonstrate that effects may be negative and need not sum to full $R^2$;
- full and reduced models use identical outer-test rows and no held-out information during tuning;
- single-informative-band simulations recover positive drop-out value without promising exact `theta` recovery;
- runtime reflects the additional reduced-model fits and can be disabled by default.

### 6. Documentation, performance envelope, and release gate

Deliver:

- a task-oriented vignette using a small reproducible multi-band encoding example;
- documentation of estimands, preprocessing, penalty scaling, CV nesting, blocked time-series requirements, result schemas, and storage costs;
- a reproducible version of the #70 simulation with absolute effects and uncertainty;
- comparison guidance for ordinary ridge, `glmnet`, R's `multiridge`, and Python's `himalaya` without claiming that no R alternatives exist;
- NEWS and lifecycle documentation.

Accept when:

- examples execute during documentation checks;
- focused tests, full package tests, `R CMD check`, and hosted CI pass on the exact commit;
- rendered documentation contains no stale API or implausible output;
- performance guardrails cover critical kernels without brittle wall-clock gates on ordinary PR jobs.

## Test strategy

Every backend and execution path must satisfy the same contracts through:

- **analytic/differential tests:** direct generalized-ridge oracle; primal/SVD/dual agreement;
- **metamorphic tests:** equal-penalty ordinary-ridge reduction, feature/band permutation invariance, batch-size invariance, consistent scaling behavior;
- **leakage tests:** outer-held-out response mutation, fold-local preprocessing checks, explicit nested-CV provenance;
- **adversarial tests:** rank deficiency, highly correlated bands, suppression, constant columns/responses, zero/near-zero band weights, extreme penalties, $p<n$, and $p>n$;
- **property tests:** finite predictions, constraint satisfaction, deterministic candidate selection, stable shapes/names across randomized valid inputs;
- **performance tests:** fixed seeded fixtures with separate runtime and peak-memory baselines.

Floating-point comparisons use condition-aware absolute and relative tolerances. Exact equality is not a general numerical requirement.

## Out of scope for this epic

- HRF construction or convolution; predictors must arrive analysis-ready;
- noise-ceiling estimation or normalization;
- GPU support;
- exhaustive commonality analysis, Shapley attribution, or a generic “shared variance” scalar;
- overlapping-searchlight fitting for an estimator whose responses are conditionally separable;
- behavioral changes to `banded_ridge_da_model()`;
- automatic deletion/PCA reduction of feature spaces to satisfy memory limits.

These can become follow-up epics after the core estimator and its estimands are stable.

## Dependency order

- [x] Child 1: contract and direct reference core
- [x] Child 2: nested per-response tuning — depends on Child 1
- [x] Child 3: rMVPA model and voxelwise execution — depends on Children 1–2
- [x] Child 4: optimized SVD/dual backends — depends on Child 1; integrates after Child 2 contracts stabilize
- [x] Child 5: predictive feature-set attribution — depends on Children 2–4
- [x] Child 6: documentation and release gate — depends on Children 1–5

Child issues should link back here and close only their stated slice. Completing an early child does not imply completion of this epic.

## Epic definition of done

This epic is complete only when:

1. users can fit a single-domain, multi-band encoding model through public rMVPA APIs;
2. per-response tuning is nested, reproducible, and mutation-tested against leakage;
3. direct, SVD, and dual paths satisfy the same numerical contract;
4. whole-brain execution is chunked and emits unambiguous per-response outputs;
5. predictive feature-set attribution has explicit non-additive semantics;
6. coefficients and large diagnostics are opt-in with documented storage costs;
7. existing DA behavior remains unchanged or is protected by explicit parity tests;
8. package tests, `R CMD check`, documentation builds, and hosted CI pass on the release commit.

## References

- Nunez-Elizalde, Huth & Gallant (2019). *Voxelwise encoding models with non-spherical multivariate normal priors*. NeuroImage 197:482–492.
- Dupré la Tour, Eickenberg, Nunez-Elizalde & Gallant (2022). *Feature-space selection with banded ridge regression*. NeuroImage 264:119728.
- `himalaya`: grouped ridge and multiple-kernel ridge reference implementation for multi-target encoding.
- `multiridge`: existing R multi-penalty ridge implementation; useful ecosystem comparison, not an rMVPA integration substitute.

**Comment by bbuchsbaum** [comment]
Release certification is complete at exact commit `5529a69a10fc1429d18c4e1318daae4325f81c17`.

Evidence:
- Independent generalized-ridge oracle, randomized/adversarial fixtures, nested-CV leakage mutations, attribution adversaries, and public-path tests: 1,376 focused expectations, 0 failures.
- Installed-package coverage court: 5,400 expectations, 0 failures; trace merge succeeded; 73.42% package coverage. New method-file coverage: attribution 89.16%, tuning 89.41%, model 92.25%, core 94.25%, solvers 96.00%.
- Direct/SVD/dual prediction parity is at floating-point scale; the benchmark receipt maximum scaled error is 2.7e-12, below the 1e-10 contract.
- Final local source-tarball R CMD check: 0 errors, 0 warnings, 2 pre-existing notes. Clean-consumer install/public example and the complete pkgdown site build passed.
- Exact-SHA hosted gates all succeeded: [R CMD check](https://github.com/bbuchsbaum/rMVPA/actions/runs/33200865203), [coverage](https://github.com/bbuchsbaum/rMVPA/actions/runs/33200865180), [pkgdown](https://github.com/bbuchsbaum/rMVPA/actions/runs/33200865171), [extended tests](https://github.com/bbuchsbaum/rMVPA/actions/runs/33200865210), [capability matrix](https://github.com/bbuchsbaum/rMVPA/actions/runs/33200865200), and [performance guardrails](https://github.com/bbuchsbaum/rMVPA/actions/runs/33200865212).
- The [rendered article](https://bbuchsbaum.github.io/rMVPA/articles/Banded_Ridge_Encoding.html) is live.

The registered 30-seed reproduction of #70 did **not** support the historical superiority claim: mean absolute correlation gain was -0.047565622, empirical 95% interval [-0.06868121, -0.02957905], with pure-noise calibration near zero. That negative result is reported rather than used as correctness evidence.

Local coordination receipt: Mote epic `bd-01M12V25A892F2PDWMNZQQD6GB` and BR1 through BR6 are closed.

**Comment by bbuchsbaum** [comment]
Closing: the epic definition of done is satisfied at `5529a69a10fc1429d18c4e1318daae4325f81c17`, with all exact-commit release gates green.
